source/simulator/introspection.cc

```diff
@@ -32,166 +32,157 @@ namespace aspect
 {
   namespace internal
   {
-
     /**
-     * Return pair with @p n_components and a filled ComponentIndices structure.
+     * Return a filled ComponentIndices structure.
      */
     template <int dim>
-    std::pair<unsigned int, typename Introspection<dim>::ComponentIndices>
-    setup_component_indices (const unsigned int n_compositional_fields)
+    typename Introspection<dim>::ComponentIndices
+    setup_component_indices (dealii::FEVariableSystem<dim> &ib)
     {
       typename Introspection<dim>::ComponentIndices ci;
-      unsigned int comp = 0;
-
       for (unsigned int i=0; i<dim; ++i)
-        ci.velocities[i] = comp++;
-      ci.pressure = comp++;
-      ci.temperature = comp++;
+        ci.velocities[i] = ib.variable("velocity").first_component_index+i;
+
+      ci.pressure = ib.variable("pressure").first_component_index;
+      ci.temperature = ib.variable("temperature").first_component_index;
+      unsigned int n_compositional_fields = ib.variable("compositions").n_components();
       for (unsigned int i=0; i<n_compositional_fields; ++i)
-        ci.compositional_fields.push_back(comp++);
-      return std::make_pair(comp, ci);
+        ci.compositional_fields.push_back(ib.variable("compositions").first_component_index+i);
```

Comment:
@gassmoeller suggested we could put it into a try... catch... block and set it to a signalling NAN if it's not there. Then it would be possible to remove one of the variables if it is not needed, and it would work as long as this variable is not accessed anywhere. I just thought we may want to think about if this is something we want to allow, it's nothing we necessarily have to change.